### PR TITLE
Support runtime specific configurations.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/containerd/containerd/issues/2701.

Support per-runtime configuration in the `cri` config file. The newly introduced `Options` field can be any type. Currently, only supported types are [`options.Options`](https://github.com/containerd/containerd/blob/master/runtime/v2/runc/options/oci.pb.go#L39) for shim v2, and [`runctypes.RuncOptions`](https://github.com/containerd/containerd/blob/master/runtime/linux/runctypes/runc.pb.go#L40). `runctypes.RuncOptions` is the fallback for unknown runtime types.

Change [`getRuntimeType`](https://github.com/containerd/cri/pull/943/files#diff-f41a864ccb9e17c8853f725df14b8e3eR479) to support new option types.

Here is an example config which uses runc with shim v1 as the default runtime, runc with shim v2 as the `runc` runtime, and kata as the `kata` runtime:
```toml
[plugins.cri.containerd.default_runtime]                                                                                           
  runtime_type = "io.containerd.runtime.v1.linux"
# https://github.com/containerd/containerd/blob/v1.2.0-rc.1/runtime/linux/runctypes/runc.pb.go#L40
[plugins.cri.containerd.default_runtime.options]
  Runtime = "runc"
  RuntimeRoot = "/run/containerd/default"

[plugins.cri.containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v1"
# https://github.com/containerd/containerd/blob/v1.2.0-rc.1/runtime/v2/runc/options/oci.pb.go#L39
[plugins.cri.containerd.runtimes.runc.options]
  BinaryName = "runc"
  Root = "/run/containerd/runc"

[plugins.cri.containerd.runtimes.kata]
  runtime_type = "io.containerd.runtime.kata.v2"
```

Note that the deprecated daemon level config `systemd_cgroup`, `no_pivot`, and runtime level config `runtime_engine`, `runtime_root` still work for the legacy runtime `io.containerd.runtime.v1.linux`. However, 1) they won't work for shim v2 runtimes; 2) they'll be deprecated when shim v1 is deprecated.

It is recommended to use the new `Options` field for all runtimes.

/cc @yujuhong @crosbymichael @tallclair 
Signed-off-by: Lantao Liu <lantaol@google.com>

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -78,16 +89,19 @@ The explanation and default value of each configuration item are as follows:
...
-
-      # runtime_root is the directory used by containerd for runtime state.
-      runtime_root = ""
+      runtime_type = "io.containerd.runc.v1"
+
+      # "plugins.cri.containerd.runtimes.runc.options" is options specific to
+      # "io.containerd.runc.v1". See:
+      #   https://github.com/containerd/containerd/blob/v1.2.0-rc.1/runtime/v2/runc/options/oci.pb.go#L39.
+      [plugins.cri.containerd.runtimes.runc.options]
+        NoPivotRoot = false
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply docs/config.md]: All of them are supported now actually. We just pass through whatever configurations from the config file.

[the remark this answers is not in the document]
Comment (mikebrow) [reply docs/config.md]: Granted these .md's could use some more content but probably should document the fields here vs say go read the code :-)

[the remark this answers is not in the document]
Comment (Random-Liu) [reply docs/config.md]: Will document them then. :)

[the remark this answers is not in the document]
Comment (mikebrow) [reply docs/config.md]: criu supported ? FYI that's checkpoint and restore..

[the remark this answers is not in the document]
Comment (Random-Liu) [reply docs/config.md]: I mean the option is supported, means that you can config it. We are just not using it.

[the remark this answers is not in the document]
Comment (Random-Liu) [reply docs/config.md]: We now support whatever configurations supported by the runtime, we directly pass through the config, and don't do anything in the cri plugin.

[the remark this answers is not in the document]
Comment (mikebrow) [reply docs/config.md]: A little surprised as that's a different tact than we've taken in the past, but I like that..good! With that in mind, I think it's worth mentioning to the user that while the field is there, kubernetes does not have a feature as yet to take advantage, then if there is an open issue in kubernetes we could link to it. Mainly I was looking at checkpoint/restore going from TODO to configured, but not enabled.

File: pkg/server/helpers.go [hunk 2]
```diff
@@ -464,3 +454,45 @@ func parseImageReferences(refs []string) ([]string, []string) {
...
+			Runtime:       r.Engine,
+			RuntimeRoot:   r.Root,
+			SystemdCgroup: c.SystemdCgroup,
+		}, nil
+	}
+	options := getRuntimeType(r.Type)
+	if err := toml.PrimitiveDecode(*r.Options, options); err != nil {
+		return nil, err
+	}
+	return options, nil
```
[the remark this answers is not in the document]
Comment (mikebrow) [reply pkg/server/helpers.go]: ok... fail not needed, pass through new meta :-)

File: docs/config.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply docs/config.md]: Yeah, our goal is to support multiple runtimes now, :) If not necessary, we may not want to add runtime specific logic.

**Pull request merged**